ctapipe/reco/tests/test_HillasReconstructor.py

```diff
@@ -197,29 +361,104 @@ def test_invalid_events():
                 geom, dl1.image, picture_thresh=10.0, boundary_thresh=5.0
             )
 
+            dl1.parameters = ImageParametersContainer()
+
             try:
                 moments = hillas_parameters(geom[mask], dl1.image[mask])
                 hillas_dict[tel_id] = moments
+                dl1.parameters.hillas = moments
             except HillasParameterizationError:
+                dl1.parameters.hillas = HillasParametersContainer()
                 continue
 
-        # construct a dict only containing the last telescope events
-        # (#telescopes < 2)
-        hillas_dict_only_one_tel = dict()
-        hillas_dict_only_one_tel[tel_id] = hillas_dict[tel_id]
-        with pytest.raises(TooFewTelescopesException):
-            fit.predict(hillas_dict_only_one_tel, subarray, tel_azimuth, tel_altitude)
-
-        # construct a hillas dict with the width of the last event set to 0
-        # (any width == 0)
-        hillas_dict_zero_width = hillas_dict.copy()
-        hillas_dict_zero_width[tel_id]["width"] = 0 * u.m
-        with pytest.raises(InvalidWidthException):
-            fit.predict(hillas_dict_zero_width, subarray, tel_azimuth, tel_altitude)
-
-        # construct a hillas dict with the width of the last event set to np.nan
-        # (any width == nan)
-        hillas_dict_nan_width = hillas_dict.copy()
-        hillas_dict_zero_width[tel_id]["width"] = np.nan * u.m
-        with pytest.raises(InvalidWidthException):
-            fit.predict(hillas_dict_nan_width, subarray, tel_azimuth, tel_altitude)
+        # copy event container to modify it
+        event_copy = deepcopy(event)
+        # overwrite all image parameters but the last one with dummy ones
+        for tel_id in list(event_copy.dl1.tel.keys())[:-1]:
+            event_copy.dl1.tel[tel_id].parameters.hillas = HillasParametersContainer()
+        fit(event_copy)
+        assert event_copy.dl2.stereo.geometry["HillasReconstructor"].is_valid is False
+
+        # Now use the original event, but overwrite the last width to 0
+        event.dl1.tel[tel_id].parameters.hillas.width = 0 * u.m
+        fit(event)
+        assert event.dl2.stereo.geometry["HillasReconstructor"].is_valid is False
+
+        # Now use the original event, but overwrite the last width to NaN
+        event.dl1.tel[tel_id].parameters.hillas.width = np.nan * u.m
+        fit(event)
+        assert event.dl2.stereo.geometry["HillasReconstructor"].is_valid is False
+
+
+def test_reconstruction_against_simulation(subarray_and_event_gamma_off_axis_500_gev):
```

Comment:
These tests contain much code to setup the initial events / hillas parameterization. 

Please move those to fixtures, so that the code in the test body actually is the important part that is being tested.

Also, this should remove duplication between the tests.

Reply:
which tests in particular? they don't do the same things with the same files and settings

Reply:
not really what you meant I guess, but I think it's better this way: I merged the 2 tests which do the same thing (CameraFrame against TelescopeFrame), while the one TelescopeFrame against simulation does things slightly different also using a different fixture, so I left it as it is.